File: apps/burner-windows/tests/VMInventoryStoreTests.cs
using System;
using System.IO;
using System.Linq;
using Xunit;
using Flagship.Burner.VM;

namespace Flagship.Burner.Tests;

/// <summary>Direct port of apps/burner-mac VMInventoryStoreTests.swift
/// (Windows-native layout: disk.qcow2 + efi-vars.fd).</summary>
public sealed class VMInventoryStoreTests : IDisposable
{
    private readonly string _root;
    private readonly VMInventoryStore _store;

    public VMInventoryStoreTests()
    {
        _root = Path.Combine(Path.GetTempPath(), $"vm-store-tests-{Guid.NewGuid()}");
        Directory.CreateDirectory(_root);
        _store = new VMInventoryStore(new VMBundleLayout(_root));
    }

    public void Dispose()
    {
        try { Directory.Delete(_root, recursive: true); } catch { }
    }

    private static VMRecord Record(string name) => new()
    {
        Config = new VMConfig
        {
            Name = name,
            ServerDomain = name,
            Username = "harry",
            ServerName = name.Split('.').FirstOrDefault() ?? "srv",
            CpuCount = 4,
            MemoryBytes = 6 * VMResourcePlan.GiB,
            MainDiskSizeBytes = VMResourcePlan.DefaultMainDiskSizeBytes,
            NetworkMode = VMNetworkMode.Nat,
            SerialConsoleEnabled = false,
            BootUnlockMode = "auto",
            DiskEncrypted = true,
        },
        State = VMState.Created,
        CreatedAt = DateTimeOffset.FromUnixTimeSeconds(1_750_000_000),
        Tier = ServerTier.HostedVM,
    };

    // ---- Layout ----

    [Fact]
    public void BundleLayoutPaths()
    {
        var layout = new VMBundleLayout(_root);
        Assert.Equal(Path.Combine(_root, "home.harry.flagship.services"),
                     layout.BundleDir("home.harry.flagship.services"));
        Assert.Equal("config.json", Path.GetFileName(layout.ConfigPath("a.b")));
        Assert.Equal("disk.qcow2", Path.GetFileName(layout.DiskImagePath("a.b")));
        Assert.Equal("installer.iso", Path.GetFileName(layout.InstallerIsoPath("a.b")));
        Assert.Equal("efi-vars.fd", Path.GetFileName(layout.EfiVariableStorePath("a.b")));
        Assert.Equal("console.log", Path.GetFileName(layout.ConsoleLogPath("a.b")));
    }

    // ---- CRUD ----

    [Fact]
    public void CreateLoadRoundTrip()
    {
        var rec = Record("home.harry.flagship.services");
        _store.Create(rec);
        var back = _store.Load("home.harry.flagship.services");
        Assert.Equal(rec, back);
        Assert.True(File.Exists(_store.Layout.ConfigPath(rec.Config.Name)));
    }

    [Fact]
    public void CreateRefusesToClobber()
    {
        var rec = Record("home.harry.flagship.services");
        _store.Create(rec);
        var ex = Assert.Throws<VMStoreException>(() => _store.Create(rec));
        Assert.Equal(VMStoreErrorKind.AlreadyExists, ex.Kind);
        Assert.Equal("home.harry.flagship.services", ex.Name);
    }

    [Fact]
    public void SavePersistsAStateChange()
    {
        var rec = Record("home.harry.flagship.services");
        _store.Create(rec);
        _store.Save(rec with { State = VMState.AwaitingPhoneUnlock });
        Assert.Equal(VMState.AwaitingPhoneUnlock, _store.Load(rec.Config.Name).State);
    }

    [Fact]
    public void SaveWithoutCreateFails()
    {
        var ex = Assert.Throws<VMStoreException>(() => _store.Save(Record("ghost.x.flagship.services")));
        Assert.Equal(VMStoreErrorKind.NotFound, ex.Kind);
    }

    [Fact]
    public void FailedStateRoundTripsWithItsReason()
    {
        var rec = Record("home.harry.flagship.services");
        _store.Create(rec);
        var failed = VMState.Failed(VMFailurePhase.Install, "stopped after 12s");
        _store.Save(rec with { State = failed });
        Assert.Equal(failed, _store.Load(rec.Config.Name).State);
    }

    [Fact]
    public void DeleteRemovesTheWholeBundle()
    {
        var rec = Record("home.harry.flagship.services");
        _store.Create(rec);
        // Simulate a disk image sitting in the bundle.
        File.WriteAllBytes(_store.Layout.DiskImagePath(rec.Config.Name), new byte[] { 0 });
        _store.Delete(rec.Config.Name);
        Assert.False(Directory.Exists(_store.Layout.BundleDir(rec.Config.Name)));
        Assert.Throws<VMStoreException>(() => _store.Load(rec.Config.Name));
    }

    [Fact]
    public void DeleteMissingFails()
    {
        var ex = Assert.Throws<VMStoreException>(() => _store.Delete("nope.flagship.services"));
        Assert.Equal(VMStoreErrorKind.NotFound, ex.Kind);
    }

    // ---- Listing (multi-server per spec) ----

    [Fact]
    public void ListReturnsAllRecordsSortedByName()
    {
        _store.Create(Record("b.bob.flagship.services"));
        _store.Create(Record("a.alice.flagship.services"));
        var names = _store.List().Select(r => r.Config.Name).ToArray();
        // Different owners on one machine is a supported posture.
        Assert.Equal(new[] { "a.alice.flagship.services", "b.bob.flagship.services" }, names);
    }

    [Fact]
    public void ListSkipsCorruptEntriesWithoutFailingTheRest()
    {
        _store.Create(Record("good.harry.flagship.services"));
        var badDir = Path.Combine(_root, "bad.harry.flagship.services");
        Directory.CreateDirectory(badDir);
        File.WriteAllText(Path.Combine(badDir, "config.json"), "not json");
        var names = _store.List().Select(r => r.Config.Name).ToArray();
        Assert.Equal(new[] { "good.harry.flagship.services" }, names);
    }

    [Fact]
    public void ListOnEmptyOrMissingRootIsEmpty()
    {
        Assert.Empty(_store.List());
        var missing = new VMInventoryStore(new VMBundleLayout(Path.Combine(_root, "does-not-exist")));
        Assert.Empty(missing.List());
    }

    // ---- Name validation ----

    [Fact]
    public void HostileNamesAreRejected()
    {
        foreach (var bad in new[] { "", ".", "..", "../escape", "..\\escape", "a/b", "a\\b", ".hidden", "UPPER.case", "spa ce", "trailing." })
        {
            var rec = Record("ok.flagship.services");
            rec = rec with { Config = rec.Config with { Name = bad, ServerDomain = bad } };
            var ex = Assert.Throws<VMStoreException>(() => _store.Create(rec));
            Assert.Equal(VMStoreErrorKind.InvalidName, ex.Kind);
        }
    }
}

File: apps/builder-mac/Tests/FlagshipBuilderTests/VMInventoryStoreTests.swift
import XCTest
@testable import FlagshipBuilderCore

final class VMInventoryStoreTests: XCTestCase {
    private var root: URL!
    private var store: VMInventoryStore!

    override func setUpWithError() throws {
        root = URL(fileURLWithPath: NSTemporaryDirectory())
            .appendingPathComponent("vm-store-tests-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: root, withIntermediateDirectories: true)
        store = VMInventoryStore(layout: VMBundleLayout(root: root))
    }

    override func tearDownWithError() throws {
        try? FileManager.default.removeItem(at: root)
    }

    private func record(name: String) -> VMRecord {
        let cfg = VMConfig(name: name,
                           serverDomain: name,
                           username: "harry",
                           serverName: String(name.split(separator: ".").first ?? "srv"),
                           cpuCount: 4,
                           memoryBytes: 6 * VMResourcePlan.gib,
                           mainDiskSizeBytes: VMResourcePlan.defaultMainDiskSizeBytes,
                           networkMode: .nat,
                           serialConsoleEnabled: false,
                           bootUnlockMode: "auto",
                           diskEncrypted: true)
        return VMRecord(config: cfg,
                        state: .created,
                        createdAt: Date(timeIntervalSince1970: 1_750_000_000),
                        tier: .hostedVM)
    }

    // MARK: - Layout

    func testBundleLayoutPaths() {
        let layout = VMBundleLayout(root: root)
        let dir = layout.bundleDir("home.harry.flagship.services")
        XCTAssertEqual(dir, root.appendingPathComponent("home.harry.flagship.services", isDirectory: true))
        XCTAssertEqual(layout.configURL("a.b").lastPathComponent, "config.json")
        XCTAssertEqual(layout.diskImageURL("a.b").lastPathComponent, "disk.img")
        XCTAssertEqual(layout.installerISOURL("a.b").lastPathComponent, "installer.iso")
        XCTAssertEqual(layout.efiVariableStoreURL("a.b").lastPathComponent, "efi-vars.bin")
        XCTAssertEqual(layout.consoleLogURL("a.b").lastPathComponent, "console.log")
    }

    // MARK: - CRUD

    func testCreateLoadRoundTrip() throws {
        let rec = record(name: "home.harry.flagship.services")
        try store.create(rec)
        let back = try store.load(name: "home.harry.flagship.services")
        XCTAssertEqual(back, rec)
        XCTAssertTrue(FileManager.default.fileExists(
            atPath: store.layout.configURL(rec.config.name).path))
    }

    func testLegacyRecordWithoutActivityTimestampsStillLoads() throws {
        let rec = record(name: "legacy.harry.flagship.services")
        try store.create(rec)
        let url = store.layout.configURL(rec.config.name)
        var json = try JSONSerialization.jsonObject(with: Data(contentsOf: url)) as! [String: Any]
        json.removeValue(forKey: "stateChangedAt")
        json.removeValue(forKey: "lastConnectedAt")
        try JSONSerialization.data(withJSONObject: json).write(to: url)

        let loaded = try store.load(name: rec.config.name)
        XCTAssertNil(loaded.stateChangedAt)
        XCTAssertNil(loaded.lastConnectedAt)
    }

    func testCreateRefusesToClobber() throws {
        let rec = record(name: "home.harry.flagship.services")
        try store.create(rec)
        XCTAssertThrowsError(try store.create(rec)) { err in
            XCTAssertEqual(err as? VMStoreError, .alreadyExists("home.harry.flagship.services"))
        }
    }

    func testSavePersistsAStateChange() throws {
        var rec = record(name: "home.harry.flagship.services")
        try store.create(rec)
        rec.state = .awaitingPhoneUnlock
        try store.save(rec)
        XCTAssertEqual(try store.load(name: rec.config.name).state, .awaitingPhoneUnlock)
    }

    func testSaveWithoutCreateFails() {
        XCTAssertThrowsError(try store.save(record(name: "ghost.x.flagship.services"))) { err in
            XCTAssertEqual(err as? VMStoreError, .notFound("ghost.x.flagship.services"))
        }
    }

    func testDeleteRemovesTheWholeBundle() throws {
        let rec = record(name: "home.harry.flagship.services")
        try store.create(rec)
        // Simulate a disk image sitting in the bundle.
        FileManager.default.createFile(
            atPath: store.layout.diskImageURL(rec.config.name).path, contents: Data([0]))
        try store.delete(name: rec.config.name)
        XCTAssertFalse(FileManager.default.fileExists(
            atPath: store.layout.bundleDir(rec.config.name).path))
        XCTAssertThrowsError(try store.load(name: rec.config.name))
    }

    func testDeleteMissingFails() {
        XCTAssertThrowsError(try store.delete(name: "nope.flagship.services")) { err in
            XCTAssertEqual(err as? VMStoreError, .notFound("nope.flagship.services"))
        }
    }

    // MARK: - Listing (multi-server per spec)

    func testListReturnsAllRecordsSortedByName() throws {
        try store.create(record(name: "b.bob.flagship.services"))
        try store.create(record(name: "a.alice.flagship.services"))
        let names = store.list().map { $0.config.name }
        // Different owners on one machine is a supported posture.
        XCTAssertEqual(names, ["a.alice.flagship.services", "b.bob.flagship.services"])
    }

    func testListSkipsCorruptEntriesWithoutFailingTheRest() throws {
        try store.create(record(name: "good.harry.flagship.services"))
        let badDir = root.appendingPathComponent("bad.harry.flagship.services", isDirectory: true)
        try FileManager.default.createDirectory(at: badDir, withIntermediateDirectories: true)
        try Data("not json".utf8).write(to: badDir.appendingPathComponent("config.json"))
        let names = store.list().map { $0.config.name }
        XCTAssertEqual(names, ["good.harry.flagship.services"])
    }

    func testListOnEmptyOrMissingRootIsEmpty() {
        XCTAssertEqual(store.list(), [])
        let missing = VMInventoryStore(layout: VMBundleLayout(
            root: root.appendingPathComponent("does-not-exist")))
        XCTAssertEqual(missing.list(), [])
    }

    // MARK: - Name validation

    func testHostileNamesAreRejected() {
        for bad in ["", ".", "..", "../escape", "a/b", ".hidden", "UPPER.case", "spa ce"] {
            var rec = record(name: "ok.flagship.services")
            rec = VMRecord(config: VMConfig(name: bad,
                                            serverDomain: bad,
                                            username: "u",
                                            serverName: "s",
                                            cpuCount: 2,
                                            memoryBytes: 1,
                                            mainDiskSizeBytes: 1,
                                            networkMode: .nat,
                                            serialConsoleEnabled: false,
                                            bootUnlockMode: "auto",
                                            diskEncrypted: true),
                           state: rec.state, createdAt: rec.createdAt, tier: rec.tier)
            XCTAssertThrowsError(try store.create(rec), "'\(bad)' must be rejected")
        }
    }
}
